File: src/stores/rule.ts
import { defineStore } from 'pinia';
import { ref } from 'vue';
import { invoke } from '@tauri-apps/api/core';
import type { Rule } from '@/types';

export const useRuleStore = defineStore('rule', () => {
  const rules = ref<Rule[]>([]);
  const isLoading = ref(false);
  const error = ref<string | null>(null);

  async function fetchRules(softwareId: string = '') {
    try {
      isLoading.value = true;
      error.value = null;
      const result = await invoke<Rule[]>('get_rules', { softwareId });
      rules.value = result;
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to fetch rules';
    } finally {
      isLoading.value = false;
    }
  }

  async function createRule(
    softwareId: string,
    name: string,
    ruleType: string,
    filePath: string,
    content: string = ''
  ) {
    try {
      isLoading.value = true;
      error.value = null;
      const now = new Date().toISOString();
      const rule: Rule = {
        id: crypto.randomUUID(),
        softwareId,
        name,
        type: ruleType,
        filePath,
        content,
        isActive: true,
        createdAt: now,
        updatedAt: now,
      };
      await invoke('create_rule', { rule });
      rules.value.push(rule);
      return rule;
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to create rule';
      throw e;
    } finally {
      isLoading.value = false;
    }
  }

  async function updateRule(rule: Rule) {
    try {
      isLoading.value = true;
      error.value = null;
      rule.updatedAt = new Date().toISOString();
      await invoke('update_rule', { rule });
      const index = rules.value.findIndex(r => r.id === rule.id);
      if (index !== -1) {
        rules.value[index] = rule;
      }
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to update rule';
      throw e;
    } finally {
      isLoading.value = false;
    }
  }

  async function deleteRule(ruleId: string) {
    try {
      isLoading.value = true;
      error.value = null;
      await invoke('delete_rule', { ruleId });
      rules.value = rules.value.filter(r => r.id !== ruleId);
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to delete rule';
      throw e;
    } finally {
      isLoading.value = false;
    }
  }

  async function toggleRule(ruleId: string, isActive: boolean) {
    try {
      isLoading.value = true;
      error.value = null;
      await invoke('toggle_rule', { ruleId, isActive });
      const rule = rules.value.find(r => r.id === ruleId);
      if (rule) {
        rule.isActive = isActive;
      }
    } catch (e) {
      error.value = e instanceof Error ? e.message : 'Failed to toggle rule';
      throw e;
    } finally {
      isLoading.value = false;
    }
  }

  function getRulesBySoftware(softwareId: string): Rule[] {
    return rules.value.filter(r => r.softwareId === softwareId);
  }

  function getRulesByType(type: string): Rule[] {
    return rules.value.filter(r => r.type === type);
  }

  return {
    rules,
    isLoading,
    error,
    fetchRules,
    createRule,
    updateRule,
    deleteRule,
    toggleRule,
    getRulesBySoftware,
    getRulesByType,
  };
});

File: src-tauri/src/commands/rule.rs
use crate::models::Rule;
use crate::utils::now_rfc3339;
use crate::AppState;
use tauri::State;

#[tauri::command]
pub fn get_rules(state: State<AppState>, software_id: String) -> Result<Vec<Rule>, String> {
    log::info!("Getting rules for software: {}", software_id);
    if software_id.is_empty() {
        state.db.get_all_rules().map_err(|e| e.to_string())
    } else {
        state.db.get_rules_by_software(&software_id).map_err(|e| e.to_string())
    }
}

#[tauri::command]
pub fn create_rule(state: State<AppState>, rule: Rule) -> Result<(), String> {
    log::info!("Creating rule: {}", rule.name);
    state.db.upsert_rule(&rule).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn update_rule(state: State<AppState>, rule: Rule) -> Result<(), String> {
    log::info!("Updating rule: {}", rule.name);
    state.db.upsert_rule(&rule).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn delete_rule(state: State<AppState>, rule_id: String) -> Result<(), String> {
    log::info!("Deleting rule: {}", rule_id);
    state.db.delete_rule(&rule_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn toggle_rule(state: State<AppState>, rule_id: String, is_active: bool) -> Result<(), String> {
    log::info!("Toggling rule {} to {}", rule_id, is_active);
    let rules = state.db.get_all_rules().map_err(|e| e.to_string())?;
    if let Some(mut rule) = rules.into_iter().find(|r| r.id == rule_id) {
        rule.is_active = is_active;
        rule.updated_at = Some(now_rfc3339());
        state.db.upsert_rule(&rule).map_err(|e| e.to_string())?;
    }
    Ok(())
}

